client/session.go

```diff
@@ -563,3 +595,936 @@ func (s *session) Close() error {
+) error {
+	var (
+		retrier = xretry.NewRetrier(xretry.NewOptions().
+			BackoffFactor(2).
+			Max(3).
+			InitialBackoff(time.Second).
+			Jitter(true))
+		enqueueCh           = newEnqueueChannel()
+		peerBlocksBatchSize = s.streamBlocksBatchSize
+	)
+
+	// Consume the incoming metadata and enqueue to the ready channel
+	go func() {
+		s.streamCollectedBlocksMetadata(len(peers), ch, enqueueCh)
+		// Begin assessing the queue and how much is processed, once queue
+		// is entirely processed then we can close the enqueue channel
+		enqueueCh.closeOnAllProcessed()
+	}()
+
+	// Fetch blocks from peers as results become ready
+	peerQueues := make(peerBlocksQueues, 0, len(peers))
+	for _, peer := range peers {
+		peer := peer
+		size := peerBlocksBatchSize
+		workers := s.streamBlocksWorkers
+		drainEvery := 100 * time.Millisecond
+		queue := newPeerBlocksQueue(peer, size, drainEvery, workers, func(batch []*blocksMetadata) {
+			s.streamBlocksBatchFromPeer(retrier, peer, batch, result, enqueueCh)
+		})
+		peerQueues = append(peerQueues, queue)
+	}
+
+	var (
+		currStart, currEligible []*blocksMetadata
+		blocksMetadataQueues    []blocksMetadataQueue
+	)
+	for perPeerBlocksMetadata := range enqueueCh.get() {
+		// Filter and select which blocks to retrieve from which peers
+		if err := s.selectBlocksForSeriesFromPeerBlocksMetadata(
+			perPeerBlocksMetadata, peerQueues,
+			currStart, currEligible, blocksMetadataQueues,
+		); err != nil {
+			enqueueCh.trackProcessed(1)
+			continue
+		}
+
+		// Insert work into peer queues
+		var wg sync.WaitGroup
+		for _, peerBlocksMetadata := range perPeerBlocksMetadata {
+			if len(peerBlocksMetadata.blocks) == 0 {
+				// No blocks assigned
+				continue
+			}
+			queue := peerQueues.findQueue(peerBlocksMetadata.peer)
+			wg.Add(1)
+			queue.enqueue(peerBlocksMetadata, wg.Done)
+		}
+		// Launch asynchronous waiter to mark completion
+		go func() {
+			wg.Wait()
+			enqueueCh.trackProcessed(1)
+		}()
+	}
+
+	// Close all queues
+	peerQueues.closeAll()
+
+	return nil
+}
+
+func (s *session) streamCollectedBlocksMetadata(
+	peersLen int,
+	ch <-chan blocksMetadata,
+	enqueueCh *enqueueChannel,
+) {
+	metadata := make(map[string]*receivedBlocks)
+
+	// Receive off of metadata channel
+	for {
+		m, ok := <-ch
+		if !ok {
+			break
+		}
+
+		received, ok := metadata[m.id]
+		if !ok {
+			received = &receivedBlocks{
+				results: make([]*blocksMetadata, 0, peersLen),
+			}
+			metadata[m.id] = received
+		}
+		if received.submitted {
+			// Already submitted to enqueue channel
+			continue
+		}
+
+		received.results = append(received.results, &m)
+
+		if len(received.results) == peersLen {
+			enqueueCh.enqueue(received.results)
+			received.submitted = true
+		}
+	}
+
+	// Enqueue all unsubmitted received metadata
+	for _, received := range metadata {
+		if received.submitted {
+			continue
+		}
+		enqueueCh.enqueue(received.results)
+	}
+}
+
+func (s *session) selectBlocksForSeriesFromPeerBlocksMetadata(
```

Comment:
This was apart of another function already, I'm kind of inclined not too tbh just based on what it might look like was kinda confusing when I tried to do this.